### PR TITLE
Added explicit keyword to constructors with one argument

### PR DESCRIPTION
**Pull request by syclik**
#### Summary:

Adds the `explicit` keyword to constructors with one argument.

#### Intended Effect:

Code cleanup. Matches closer to #1361.

#### How to Verify:

Visually inspect the code for correctness.

#### Side Effects:

I don't foresee any problems, but may trickle upstream to CmdStan.

#### Documentation:

None required.

#### Reviewer Suggestions: 

@betanalpha

### CODE REVIEW

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by syclik** [comment]
Jenkins, retest this please.

File: src/stan/services/arguments/singleton_argument.hpp [hunk 1]
```diff
@@ -44,7 +44,7 @@ namespace stan {
     template<typename T>
     class singleton_argument: public valued_argument {
     public:
-      singleton_argument(): _validity("All") {
+      explicit singleton_argument(): _validity("All") {
```
Comment (bob-carpenter) [inline src/stan/services/arguments/singleton_argument.hpp]: Why are you making this unary constructor explicit?  Will this still work:

```
singleton_argument a;
```

?

File: src/stan/services/arguments/argument.hpp [hunk 2]
```diff
@@ -15,7 +15,7 @@ namespace stan {
         : indent_width(2),
           help_width(20) { }
 
-      argument(const std::string name)
+      explicit argument(const std::string name)
```
Comment (bob-carpenter) [inline src/stan/services/arguments/argument.hpp]: Shouldn't this be taking a constant reference, not copying the string?

**Comment by syclik** [comment]
Bob, thanks for reviewing the code! My fault: I did the laziest thing possible. I copied code over from the other branch so squash another 4 file diffs.

I think it's clear that I shouldn't have done this. Here are some options now:
1. close this pull request and fix Michael's branch to match `develop`
2. walk through all of the arguments and fix this permanently on the `develop` branch
3. walk through just these changes and fix it for this pull request

I would feel most comfortable with 2, but 1 is the easiest thing to do. Thoughts?

**Comment by bob-carpenter** [comment]
I think (2) with as much of (3) as you have time for is good.
No reason to put off changes that move in the right direction just
because there are more waiting in the wings.

This is a direct consequence of the cpplint test because the inclination
every time I see them spit out warnings is to fix them.

- Bob

**Comment by betanalpha** [comment]
I don’t remember touching the arguments in the originating
pull request so I’m not sure where this came up, or why the
warnings wouldn’t be in development.

**Comment by syclik** [comment]
It was part of 136a1a2 "Fix a swarm of cpplint errors". No action required.
If you want to help with this particular thing, just make the right fixes
to develop in a branch and we'll get it merged in. It was just cluttering
up the file diff for that pull request, so I'm trying to knock it down to
something manageable. It's down to 119 now! I'm still really hopeful.

File: src/stan/services/arguments/singleton_argument.hpp (hunk 1 above)
Comment (syclik) [inline src/stan/services/arguments/singleton_argument.hpp]: My bad. It should have gone on the next constructor. I'm fixing it.

**Comment by syclik** [comment]
I updated the branch. @betanalpha, want to take a look?

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by betanalpha** [comment]
Passing the strings by reference isn't going to help much here as the calls almost always are done with a string that gets constructed on the fly anyways (pretty sure the compiler can optimize this to a single constructor), and they don't get called enough for it really matter.  But it's largely irrelevant.

:+1:

**Comment by syclik** [comment]
Agreed. I was reading the stack overflow threads about passing string by
value being optimized for c++11, but it looks like the right thing to do in
most cases is pass by reference.

Either way, this won't affect runtime noticeably, but I'm doing this as a
way to clean up our code base and to reduce the unnecessary diffs on that
large pull request.

I'll merge when tests pass.

**Pull request merged**